### PR TITLE
batt_smbus: fix potential buffer overrun bug

### PR DESCRIPTION
**Pull request by dakejahl**
The smbus block_read command will always return the `data[]` + 2 bytes of address information. This was not being respected for the `dataflash_write` and `manufacturer_read` functions. These functions now expect the caller to provide the correctly sized buffer for the block_read command.

@themcfadden please verify on your end, thanks!

### CODE REVIEW

File: src/drivers/batt_smbus/batt_smbus.cpp [hunk 1]
```diff
@@ -490,8 +490,15 @@ int BATT_SMBUS::manufacturer_read(const uint16_t cmd_code, void *data, const uns
 	}
 
 	// returns the 2 bytes of addr + data[]
-	result = _interface->block_read(code, data, length + 2, true);
-	memcpy(data, &((uint8_t *)data)[2], length);
+	size_t len = length + 2;
+	uint8_t buf[len] = {};
```
Comment (bkueng) [inline src/drivers/batt_smbus/batt_smbus.cpp]: Please don't use variable array sizes on the stack. Rather make sure the caller provides a large enough array.

File: src/drivers/batt_smbus/batt_smbus.cpp [hunk 2]
```diff
@@ -364,7 +358,7 @@ int BATT_SMBUS::dataflash_write(uint16_t &address, void *data, const unsigned le
 {
 	uint8_t code = BATT_SMBUS_MANUFACTURER_BLOCK_ACCESS;
 
-	uint8_t tx_buf[DATA_BUFFER_SIZE + 2] = {};
+	uint8_t tx_buf[MAC_DATA_BUFFER_SIZE + 2] = {};
 
 	tx_buf[0] = ((uint8_t *)&address)[0];
 	tx_buf[1] = ((uint8_t *)&address)[1];
```
Comment (bkueng) [inline src/drivers/batt_smbus/batt_smbus.cpp]: The line below (`memcpy(&tx_buf[2], data, length);`) can also produce an overflow if `length > MAC_DATA_BUFFER_SIZE`. Can you check for that?

File: src/drivers/batt_smbus/batt_smbus.cpp [hunk 3]
```diff
@@ -489,9 +483,8 @@ int BATT_SMBUS::manufacturer_read(const uint16_t cmd_code, void *data, const uns
 		return result;
 	}
 
-	// returns the 2 bytes of addr + data[]
-	result = _interface->block_read(code, data, length + 2, true);
-	memcpy(data, &((uint8_t *)data)[2], length);
+	result = _interface->block_read(code, data, length, true);
+	memcpy(data, &((uint8_t *)data)[2], length - 2); // remove the address bytes
```
Comment (bkueng) [inline src/drivers/batt_smbus/batt_smbus.cpp]: You need to use `memmove` here.

File: src/lib/drivers/smbus/SMBus.hpp [hunk 4]
```diff
@@ -62,8 +62,8 @@ class SMBus : public device::I2C
 	/**
 	 * @brief Sends a block read command.
 	 * @param cmd_code The command code.
-	 * @param data The returned data.
-	 * @param length The number of bytes being read
+	 * @param data The returned data. The returned data will always contain the data[] + 2 bytes of address information.
```
Comment (bkueng) [inline src/lib/drivers/smbus/SMBus.hpp]: Can you clarify that the 2 additional bytes are in front?

**Comment by themcfadden** [comment]
I've been updating this  driver to work with the BQ40Z80. There was another buffer bug. Let me dig it up...

**Comment by LorenzMeier** [comment]
@dakejahl Could you please follow up?

**Comment by dakejahl** [comment]
@themcfadden are you sure? The technical reference specifies 21 bytes being returned, could you explain?

File: src/drivers/batt_smbus/batt_smbus.cpp (hunk 3 above)
Comment (dakejahl) [reply src/drivers/batt_smbus/batt_smbus.cpp]: thanks, learned something new today!

Comment (dakejahl) [reply src/drivers/batt_smbus/batt_smbus.cpp]: thanks, I learned something new today!

File: src/drivers/batt_smbus/batt_smbus.cpp [hunk 5]
```diff
@@ -339,23 +339,17 @@ void BATT_SMBUS::set_undervoltage_protection(float average_current)
...
 		return result;
 	}
 
-	// @NOTE: The data buffer MUST be 32 bytes.
-	result = _interface->block_read(code, data, DATA_BUFFER_SIZE + 2, true);
-
-	// When reading a BATT_SMBUS_MANUFACTURER_BLOCK_ACCESS the first 2 bytes will be the command code
-	// We will remove these since we do not care about the command code.
-	//memcpy(data, &((uint8_t *)data)[2], DATA_BUFFER_SIZE);
+	result = _interface->block_read(code, data, length, true);
```
Comment (mcsauder) [inline src/drivers/batt_smbus/batt_smbus.cpp]: @dakejahl , you could drop one LOC by inverting the logic:
```
if (result == PX4_OK) {
    result = _interface->block_read(code, data, length, true);
}

return result;

**Comment by mcsauder** [comment]
@themcfadden , you might be correct, I haven't dug deeply enough, but I don't see how that affects this PR.

**Pull request merged**